feat: add obligations (#150)
opentdf/docs: docs/components/policy/obligations.md

# Obligations

> Obligations are policy constructs that enable conditional access enforcement at the Policy Enforcement Point (PEP) level. They define additional restrictions and requirements beyond basic attribute-based access control (ABAC), allowing security administrators to enforce conditions such as multi-factor authentication (MFA), watermarking, encryption, or time-based access controls.

Reviewer: ```suggestion
Obligations are policy constructs that enable Policy Decision Point(PDP) - to - Policy Enforcement Point (PEP) directives that accompany a decision. They express additional controls the PEP should enforce—such as requiring MFA, applying watermarking, encrypting content, or limiting access by time. In effect, the PDP is saying “permit, provided these controls are enforced.” The PDP cannot compel or verify enforcement; it relies on the PEP to honor the obligations. 
```

> 

Reviewer: ```suggestion
If the PEP cannot or will not enforce an obligation, it should decline to grant access.
```
Does this make sense to add?

```mermaid
graph LR;
Namespace_A-->Definition_A;
Definition_A-->Value_A;
Definition_A-->Value_B;

Namespace_A-->Definition_B;
Definition_B-->Value_C;
Definition_B-->Value_D;
```

## Standard Obligations

Standard Obligations are enforced by the Access PDP (Policy Decision Point) when validating whether an Entity of a specified Category can perform an Action on a given Resource. These obligations must be satisfied for access to be granted.

A Standard Obligation includes:
- **Entity Category**: The type of entity (SUBJECT or ENVIRONMENT) that must satisfy this obligation
- **Action**: The specific action being performed (READ, STORE, CREATE, UPDATE, DELETE)
- **Resource Attribute Value**: The data attribute for which this obligation is scoped
- **Obligation Value**: The required obligation value that must be satisfied

### Examples of Standard Obligations:

#### Watermarking
Ensure document traceability:
- **Obligation**: `https://example.com/obl/drm/value/watermarking`
- **Applied to**: sensitive documents
- **Condition**: PEP must apply watermark

#### Prevent Download

> Ensure persistent file protection (PFP):

Reviewer: ```suggestion
Prevent download of plaintext content:
```
PFP is a loaded term, let's keep it simple

> - **Obligation**: `https://example.com/obl/drm/value/pfp`

Reviewer: ```suggestion
- **Obligation**: `https://example.com/obl/drm/value/no-download`
```

# Obligations

Obligations are policy constructs that enable Policy Decision Point(PDP) - to - Policy Enforcement Point (PEP) directives that accompany a decision. They express additional controls the PEP should enforce—such as requiring MFA, applying watermarking, encrypting content, or limiting access by time. In effect, the PDP is saying “permit, provided these controls are enforced.” The PDP cannot compel or verify enforcement; it relies on the PEP to honor the obligations. 
If the PEP cannot or will not enforce an obligation, it should decline to grant access.
## Composition

An obligation consists of:

1. A Namespace
2. A Definition
3. Values
4. Triggers

Platform Policy Obligations can contain multiple Namespaces, each with multiple Definitions, and each Definition can have multiple Values. Each Value can have multiple Triggers. Each trigger can have an Action, Attribute Value, and PEP identifier.

```mermaid

> graph LR;

Reviewer 2: Can we add a case of a global obligation that is triggered by an action/attribute value but _no_ PEP? For example, an obligation to validate a specific assertion could be global across all PEPs.

Author: added with explanation and updated diagram example

%% Every definition is namespaced.
Namespace_A-->ObligationDefinition_A;

%% Define definition and value A.
ObligationDefinition_A-->ObligationValue_A;

%% Define triggers.
ObligationValue_A-->Trigger_A1
Trigger_A1-->Action_A1
Trigger_A1-->AttributeValue_A1
Trigger_A1-->PEP_A1

ObligationValue_A-->Trigger_A2
Trigger_A2-->Action_A2
Trigger_A2-->AttributeValue_A2
Trigger_A2-->PEP_A2

%% Define definition and value B.
ObligationDefinition_A-->ObligationValue_B;

%% Define triggers.
ObligationValue_B-->Trigger_B1
Trigger_B1-->Action_B1
Trigger_B1-->AttributeValue_B1
Trigger_B1-->PEP_B1

ObligationValue_B-->Trigger_B2
Trigger_B2-->Action_B2
Trigger_B2-->AttributeValue_B2
Trigger_B2-->PEP_B2
```

## Standard Obligations

Standard Obligations are enforced by the Access PDP (Policy Decision Point) when validating whether an Entity of a specified Category can perform an Action on a given Resource. These obligations must be satisfied for access to be granted.

A Standard Obligation includes:

> - **Entity Category**: The type of entity (SUBJECT or ENVIRONMENT) that must satisfy this obligation

Reviewer 2: ```suggestion
```
